### PR TITLE
Convert syncing code to TypeScript

### PR DESCRIPTION
**Pull request by nwalters512**
All these files were already type-checked, so this was a pretty straightforward conversion.

### CODE REVIEW

File: apps/prairielearn/src/sync/fromDisk/assessments.ts
Comment (nwalters512) [inline apps/prairielearn/src/sync/fromDisk/assessments.ts:1]: This file had the most changes. They were necessary to account for the fact that in TypeScript, it's much easier to construct an object all at once than to construct a barebones object and then tack properties onto it as you go.

File: apps/prairielearn/src/sprocs/sync_course_instances.sql [hunk 1]
```diff
@@ -90,7 +90,7 @@ BEGIN
         RETURNING dest.short_name AS src_short_name, dest.id AS inserted_dest_id
     )
     -- Make a map from CIID to ID to return to the caller
-    SELECT jsonb_object_agg(src_short_name, COALESCE(dest_id, inserted_dest_id))
+    SELECT COALESCE(jsonb_object_agg(src_short_name, COALESCE(dest_id, inserted_dest_id)), '{}'::JSONB)
```
Comment (nwalters512) [inline apps/prairielearn/src/sprocs/sync_course_instances.sql:93]: Zod caught the `NULL` case here!

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts [hunk 2]
```diff
@@ -60,10 +56,12 @@ export async function sync(courseId, courseData) {
   const params = [courseInstanceParams, courseId];
 
   perf.start('sproc:sync_course_instances');
-  const result = await sqldb.callOneRowAsync('sync_course_instances', params);
+  const result = await sqldb.callRow(
+    'sync_course_instances',
+    params,
+    z.record(z.string(), IdSchema),
+  );
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/sync/fromDisk/courseInstances.ts:63]:
```suggestion
  const result = await sqldb.callRow(
    'sync_course_instances',
    [courseInstanceParams, courseId],
    z.record(z.string(), IdSchema),
  );
```

This also implies removing the params in line 56.

File: apps/prairielearn/src/sync/fromDisk/questions.ts [hunk 3]
```diff
@@ -72,10 +69,8 @@ export async function sync(courseId, courseData) {
   const params = [questionParams, courseId];
 
   perf.start('sproc:sync_questions');
-  const result = await sqldb.callOneRowAsync('sync_questions', params);
+  const result = await sqldb.callRow('sync_questions', params, z.record(z.string(), IdSchema));
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/sync/fromDisk/questions.ts:72]: Same comment as above.

File: apps/prairielearn/src/sync/fromDisk/tags.ts [hunk 4]
```diff
@@ -49,19 +46,16 @@ export async function sync(courseId, courseData, questionIds) {
...
   perf.end('sproc:sync_course_tags');
 
-  /** @type {[string, any][]} */
-  const newTags = res.rows[0].new_tags_json;
+  const newTags = res.rows[0].new_tags_json as [string, any][];
   const tagIdsByName = newTags.reduce((acc, [name, id]) => {
     acc.set(name, id);
     return acc;
-  }, /** @type {Map<String, any>} */ (new Map()));
+  }, new Map<string, any>());
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/sync/fromDisk/tags.ts:53]: I know you're just changing types, but:

```suggestion
  const tagIdsByName = Object.fromEntries(newTags);
```

**Review by jonatanschroeder — APPROVED** [review]
All the comments are nits.

File: apps/prairielearn/src/sync/fromDisk/tags.ts (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/tags.ts:53]: This of course changes the type from Map to Object (Record), but this is only used in line 60 in a way where Map does not produce a significant advantage. That said, if you want to keep a Map:

```suggestion
  const tagIdsByName = new Map(newTags);
```

Comment (nwalters512) [reply apps/prairielearn/src/sync/fromDisk/tags.ts]: Changed in 6934d26.

File: apps/prairielearn/src/sync/fromDisk/questions.ts (hunk 3 above)
Comment (nwalters512) [reply apps/prairielearn/src/sync/fromDisk/questions.ts]: Done in 6934d26; I got rid of all the other separate `params` objects/arrays in that commit as well.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Done in 6934d26.

**Review by jonatanschroeder — APPROVED** [review]

**Pull request merged**